### PR TITLE
Move MVCCComputeStats implementation to C++.

### PR DESCRIPTION
**Pull request by petermattis**
This change is a work in progress. I still need to adjust or remove the
usage of ReplicaDataIterator for stats computation in
Replica.splitTrigger.

```
name                                old time/op    new time/op     delta
MVCCComputeStats1Version8Bytes-8       1.04s ± 2%      0.34s ± 1%   -67.04%  (p=0.000 n=10+10)
MVCCComputeStats1Version32Bytes-8      737ms ± 4%      242ms ± 2%   -67.23%  (p=0.000 n=10+10)
MVCCComputeStats1Version256Bytes-8     211ms ± 3%       86ms ± 4%   -59.38%  (p=0.000 n=10+10)

name                                old speed      new speed       delta
MVCCComputeStats1Version8Bytes-8    64.5MB/s ± 2%  195.7MB/s ± 1%  +203.39%  (p=0.000 n=10+10)
MVCCComputeStats1Version32Bytes-8   91.0MB/s ± 4%  277.8MB/s ± 2%  +205.12%  (p=0.000 n=10+10)
MVCCComputeStats1Version256Bytes-8   319MB/s ± 3%    785MB/s ± 5%  +146.25%  (p=0.000 n=10+10)
```

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3155)

### CODE REVIEW

**Comment by petermattis** [comment]
This is now ready for a look. I wish there was a good way to diff the Go and C++ implementations.

**Comment by tamird** [comment]
Reviewed 11 of 11 files at r1.
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

_[storage/engine/engine.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qnlw8vlZy67x_k6xM:-K3Qnlw9_FI9L41JxKKx:1421308124) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/engine.go#L76)):_
this function should take `roachpb.Key` or `mvcc.MVCCKey` rather than `[]byte`

---

_[storage/engine/mvcc_test.go, line 2548 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QoH4nLQ-6-ONBthhk:-K3QoH4nLQ-6-ONBthhl:665103079) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/mvcc_test.go#L2548)):_
would be nice to avoid overwriting `err` here

---

_[storage/engine/rocksdb/db.cc, line 1240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qs78c2pFp_bTStasX:-K3Qs78c2pFp_bTStasY:1699536011) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1240)):_
where did this value come from?

---

_[storage/engine/rocksdb/db.cc, line 1260 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QscGiy1m9KqpF6_Td:-K3QscGiy1m9KqpF6_Te:548825895) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1260)):_
s/decoded/decode/

---

_[storage/engine/rocksdb/db.cc, line 1269 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QsnhVNgRaVTJ4i9hU:-K3QsnhVNgRaVTJ4i9hV:-1908949648) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1269)):_
nit: can this not duplicate the number 12?

---

_[storage/engine/rocksdb/encoding.cc, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qrl7PWOT0Bq6X0-GV:-K3Qrl7PWOT0Bq6X0-GW:141176433) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.cc#L93)):_
would this read nicer as a loop?

---

_[storage/engine/rocksdb/encoding.cc, line 97 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrrPOWDY2k3Gm_Ihh:-K3QrrPOWDY2k3Gm_Ihi:-211174788) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.cc#L97)):_
nit: extract local `sizeof(*value)`

---

_[storage/engine/rocksdb/encoding.h, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrIpaCMyjPZEJrWuH:-K3QrIpaCMyjPZEJrWuI:-1506096058) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.h#L34)):_
did you mean `*value`? here and below

---

_[storage/engine/rocksdb_test.go, line 556 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qqu_VLi4FOw4PH4-h:-K3Qqu_VLi4FOw4PH4-i:-1542324587) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb_test.go#L556)):_
remove this

---

_[storage/replica_data_iter.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrA-LXzj7aVGvLFVh:-K3QrA-LXzj7aVGvLFVi:-110463902) ([raw file](https://github.com/cockroachdb/cockroach/blob/623d09252c5e3238b8951a6d09e7dd25905f77f1/storage/replica_data_iter.go#L70)):_
how come this could be removed?

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155)_

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

_[storage/engine/rocksdb/db.cc, line 1240 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qs78c2pFp_bTStasX:-K3Rg2SNPaz19szTXzGT:1275547647) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1240)):_
It's a constant in `mvcc.go`: `mvccVersionTimestampSize`.

---

_[storage/engine/rocksdb/db.cc, line 1260 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QscGiy1m9KqpF6_Td:-K3RPUAwhLFE94yGF0SA:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1260)):_
Done.

---

_[storage/engine/rocksdb/db.cc, line 1269 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QsnhVNgRaVTJ4i9hU:-K3Rg8R1WeFHoNrqIvhj:-1495550284) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/db.cc#L1269)):_
Yeah. Added a `FmtStatus` function which uses the `google::protobuf::StringPrintf`utility.

---

_[storage/engine/rocksdb/encoding.cc, line 93 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qrl7PWOT0Bq6X0-GV:-K3RifxBGWeOaf5ZTA1G:-843628184) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.cc#L93)):_
This mirrors the code implementation. A loop is slower.

---

_[storage/engine/rocksdb/encoding.cc, line 97 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrrPOWDY2k3Gm_Ihh:-K3RirGRyqZ7sIeXeJRf:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.cc#L97)):_
Done.

---

_[storage/engine/rocksdb/encoding.h, line 34 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrIpaCMyjPZEJrWuH:-K3RjmyYxeuXNg3yW2na:-1672005831) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb/encoding.h#L34)):_
Yes. Copy/paste error.

---

_[storage/engine/rocksdb_test.go, line 556 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qqu_VLi4FOw4PH4-h:-K3RjyUcpB9PybtxU5kJ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/rocksdb_test.go#L556)):_
Done.

---

_[storage/replica_data_iter.go, line 70 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QrA-LXzj7aVGvLFVh:-K3Rk-oorei3Sj0OWHTH:1818719315) ([raw file](https://github.com/cockroachdb/cockroach/blob/623d09252c5e3238b8951a6d09e7dd25905f77f1/storage/replica_data_iter.go#L70)):_
`ri.Seek()` already does an `advance()`. This was a no-op.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155)_

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 10 unresolved discussions.

---

_[storage/engine/engine.go, line 76 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Qnlw8vlZy67x_k6xM:-K3Rke1Ll7NvSkfyF3HM:1077939989) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/engine.go#L76)):_
All of the other methods here take a `[]byte` for keys. Feel free to change this in a subsequent PR. 

---

_[storage/engine/mvcc_test.go, line 2548 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3QoH4nLQ-6-ONBthhk:-K3Rl-Q3FQ3tsp3TqNW5:1403025779) ([raw file](https://github.com/cockroachdb/cockroach/blob/83c37394736fc55492dff4136c1265ee1c9795bb/storage/engine/mvcc_test.go#L2548)):_
I don't see why it matters, and mildly irritating to do. Certainly wouldn't improve readability. Note that we were overwriting `err` in the previous code too.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155)_

**Comment by tamird** [comment]
Reviewed 4 of 4 files at r2.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/engine/rocksdb/db.cc, line 1273 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Rll8w68g_WcvM3s_E:-K3Rll8w68g_WcvM3s_F:1713597768) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8db070709072fa0e1d00bd3f8ff69bd6fa777d/storage/engine/rocksdb/db.cc#L1273)):_
Did you want to log the key here? I don't feel strongly, but since you're using `FmtStatus`, might be nice.

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155)_

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

_[storage/engine/rocksdb/db.cc, line 1273 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155#-K3Rll8w68g_WcvM3s_E:-K3RmTm5BtzeWgtBVegA:-1247117233) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f8db070709072fa0e1d00bd3f8ff69bd6fa777d/storage/engine/rocksdb/db.cc#L1273)):_
This is using c-printf formatting which is much weaker than Go fmt. Not easy to do (unless I've forgotten something).

---


_Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3155)_

File: storage/engine/rocksdb/db.cc [hunk 1]
```diff
@@ -115,6 +120,15 @@ DBStatus ToDBStatus(const rocksdb::Status& status) {
   return ToDBString(status.ToString());
 }
 
+DBStatus FmtStatus(const char *fmt, ...) {
```
Comment (bdarnell) [inline storage/engine/rocksdb/db.cc]: Is this function covered in our tests anywhere? I didn't see any cases where we test for a failure of MVCCComputeStats, but I didn't do an exhaustive search. Would be good to have tests to make sure we get the memory management right here.

**Comment by bdarnell** [comment]
LGTM

File: storage/engine/rocksdb/db.cc (hunk 1 above)
Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: `FmtStatus` is not explicit tested, but we do have tests that exercise the code path that memory allocated by `ToDBString` is freed by `statusToError`. Do you have any thoughts on how to test that the memory was actually freed?

Comment (bdarnell) [inline storage/engine/rocksdb/db.cc]: I'm more concerned about returning a pointer to freed memory than a leak (I don't see any practical way to test for the absence of a leak). I'd just like to see a test where MVCCStats returns an error to make sure we execute FmtStatus at least once.

Comment (petermattis) [inline storage/engine/rocksdb/db.cc]: Added `TestMVCCComputeStatsError` which writes an mvcc metadata key with garbage for the value.

**Pull request merged**